### PR TITLE
pkp/pkp-lib#1815 Permit hooks to register with call priorities

### PR DESCRIPTION
**Pull request by asmecher**

### CODE REVIEW

File: classes/file/FileManager.inc.php [hunk 1]
```diff
@@ -234,7 +234,6 @@ function readFileFromPath($filePath, $output = false) {
 	function downloadFile($filePath, $mediaType = null, $inline = false, $fileName = null) {
 		$result = null;
 		if (HookRegistry::call('FileManager::downloadFile', array(&$filePath, &$mediaType, &$inline, &$result, &$fileName))) return $result;
-		$postDownloadHookList = array('FileManager::downloadFileFinished', 'UsageEventPlugin::getUsageEvent');
```
Comment (bozana) [inline classes/file/FileManager.inc.php]: I see the FileManager::downloadFileFinished is still there, at the end of the function, but what about UsageEventPlugin::getUsageEvent ?

Comment (asmecher) [inline classes/file/FileManager.inc.php]: I removed the code that clears out the registry before downloading a file; the post download hook list was used to selectively restore hooks after downloading had completed. I don't think it's a good optimization so I removed it.

File: classes/plugins/HookRegistry.inc.php [hunk 2]
```diff
@@ -61,13 +65,11 @@ static function clear($hookName) {
...
+	 * @param $hookSequence int Optional hook sequence specifier HOOK_SEQUENCE_...
 	 */
-	static function register($hookName, $callback) {
+	static function register($hookName, $callback, $hookSequence = HOOK_SEQUENCE_NORMAL) {
 		$hooks =& HookRegistry::getHooks();
-		if (!isset($hooks[$hookName])) {
-			$hooks[$hookName] = array();
-		}
-		$hooks[$hookName][] =& $callback;
+		$hooks[$hookName][$hookSequence][] =& $callback;
```
Comment (bozana) [inline classes/plugins/HookRegistry.inc.php]: Does indexes $hookName and $hookSequence exist here for sure?

Comment (asmecher) [inline classes/plugins/HookRegistry.inc.php]: As indexes in the array? No, but PHP doesn't complain in this case.

Comment (bozana) [inline classes/plugins/HookRegistry.inc.php]: Ah yes, true, sorry... :-\

**Pull request merged**